[Tang Ethan Yidong] iP

EthanYidong opened the pull request.

EthanYidong committed:
Add greeting and exit messages

EthanYidong committed:
Extract message formatting to functions

EthanYidong committed:
Extract print formatting to BunnySession class
Rename bot to Bunny

EthanYidong committed:
Implement user command echoing

Wrote a basic REPL which takes input and prints it out
Scans for input equalling “bye” to signal a quit

EthanYidong committed:
Implement adding and printing to a list of items

Created a Todo class to encapsulate the data of a Todo element
BunnySession stores an ArrayList of Todos
Updated the REPL to handle the list command

EthanYidong committed:
Implement marking and unmarking TODOs as done

Override toString() for Todo class
Add markAsDone and markAsNotDone  to Todo class

Add mark and unmark handlers to REPL

EthanYidong committed:
Format code to follow SE-EDU standards

Added spacing
Renamed constants to follow screaming snake case

EthanYidong committed:
Extract command parsing responsibility to a separate class

EthanYidong committed:
Refractor Todo to inherit from new abstract class Task

EthanYidong committed:
Refractor to move "done" functionality to the Task class.

EthanYidong committed:
Implement Deadline and Event types of Task.

Added task count to list

EthanYidong committed:
Create test case for text-based testing

EthanYidong committed:
Refractor code to delegate command validation to separate classes

Move all code to the com.ethanyidong.bunny package
Create ArgumentValidator and CommandValidator interfaces
Implement PositionalArgumentCommandValidator and FlagArgumentCommandValidator
Implement IntegerArgumentValidator, TaskIndexArgumentValidator, and NotEmptyArgumentValidator
Make commands an enum instead of a String
Implement per-command validation setup

TODO: Fix regression in which specific errors are no longer reported (e.x. "should be a number")

EthanYidong committed:
Provide helpful messages for invalid commands

Refractor Validators to throw errors for invalid inputs instead of returning false

miahjerry commented on src/main/java/Bunny.java:
```diff
@@ -0,0 +1,69 @@
+
+        bunny.printMessage("Hello! I'm Bunny.\nWhat can I do for you?");
+
+        while (true) {
```
Nice indentation. If-else blocks are clearly broken up into readable chunks.

EthanYidong committed:
Implement choosing between singular and plural nouns

Updated tests to reflect this change
Fixed issue with Windows testing

joulev commented on src/main/java/com/ethanyidong/bunny/Bunny.java:
```diff
@@ -0,0 +1,82 @@
+                continue;
+            }
+
+            switch (inputCommand.getCommand()) {
```
Really nice :) I love that you are using `switch`/`case` statements here, as it helps a lot in readability compared to `if`/`else` statements.

joulev commented on src/main/java/com/ethanyidong/bunny/BunnySession.java:
```diff
@@ -0,0 +1,43 @@
+        return this.taskList.get(index);
+    }
+
+    public int numTasks() {
```
While `numTasks` is quite clear already, I think it could be made clearer such as `getNumberOfTasks`.

joulev commented on src/main/java/com/ethanyidong/bunny/CommandType.java:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    public static CommandType fromString(String command) {
+        switch(command) {
```
```suggestion
        switch (command) {
```

and I found similar cases in some other places as well.

joulev commented on src/main/java/com/ethanyidong/bunny/Deadline.java:
```diff
@@ -0,0 +1,27 @@
+        this.by = by;
+    }
+
+    protected String label() {
```
It might help to have some comments here explaining what this label is for. Same for some other files.

joulev commented on src/main/java/com/ethanyidong/bunny/ParsedCommand.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    private static HashMap<String, String> extractFlagArguments(String[] positionalAndFlagArguments) {
+        HashMap<String, String> ret = new HashMap<>();
```
Maybe this variable should have a clearer name. `ret` is not an English word so it's not immediately clear to me what it means.

joulev commented on src/main/java/com/ethanyidong/bunny/Task.java:
```diff
@@ -0,0 +1,38 @@
+
+public abstract class Task {
+    private String name;
+    private boolean done;
```
`isDone` might be a better name in my opinion.

joulev commented on src/main/java/com/ethanyidong/bunny/BunnySession.java:
```diff
@@ -0,0 +1,43 @@
+        this.taskList = new ArrayList<>();
+    }
+
+    public void printMessage(String message) {
+        this.printMessage(Arrays.asList(message.split("\n")));
+    }
```
Seeing from the occasions this function is called, I think it might be good to use the method signature of `System.out.printf` instead, so that we can use format strings instead of a *lot* of string concatenation `+` signs.

```java
public void printMessage(String format, Object... args) {
    String message = String.format(format, args);
    this.printMessage(Arrays.asList(message.split("\n")));
}
```

joulev reviewed (commented):
Very good job, and definitely not something like an average iP repo at this time of the semester! 💯

I have a couple of suggestions and comments here, do have a look.

EthanYidong committed:
Refractor to separate command functionality into different classes

Create ExecutableCommand class
Rename ParsedCommand to TokenizedCommand for clarity
Shifted validation, parsing, and execution responsibilities to the ExecutableCommand class and its respective children

EthanYidong committed:
Fix argument passing bugs causing extra spaces to be entered into arguments

Updated text test to match

EthanYidong committed:
Cleanup code via IntelliJ

EthanYidong committed:
Delete unused code comment

EthanYidong committed:
Update README.md for token commit to create branch for Level-5

Exceptions have already been implemented and handled

EthanYidong committed:
Merge branch 'branch-Level-5'

EthanYidong committed:
Update README.md for token commit to create branch for A-Packages

Packages were already created

EthanYidong committed:
Merge branch 'branch-A-Packages'

EthanYidong committed:
Update manifest file

EthanYidong committed:
Implement deleteing tasks
Add DeleteTaskCommand
Update text-ui-test

EthanYidong committed:
Implement saving

Update text-ui-test framework to disable saving in that environment

EthanYidong committed:
Merge branch 'branch-level-6'

EthanYidong committed:
Merge branch 'branch-level-7'

sevenseasofbri commented on src/main/java/com/ethanyidong/bunny/Bunny.java:
```diff
@@ -0,0 +1,81 @@
+            if (saveEnabled) {
+                saveFileWriter.close();
+            }
+        } catch (Exception _ex) {
```
Perhaps use the standard naming convention for the Exception object in the catch block instead of `_ex`, you can just use `e` or `err`. [See here](https://nus-cs2113-ay2223s2.github.io/website/schedule/week5/topics.html#c-to-java-exceptions-how-to-use-exceptions)

sevenseasofbri commented on src/main/java/com/ethanyidong/bunny/Bunny.java:
```diff
@@ -0,0 +1,81 @@
+                }
+                reader.close();
+            } catch (Exception _ex){
+                System.out.println("Error reading save file! Continuing...");
+            } finally {
+                bunny.setIsSuppressed(false);
+            }
+        }
+
+        bunny.printMessage("Hello! I'm Bunny.\nWhat can I do for you?");
```
Avoid the usage of magic literals. You can refactor them to static final string constants. [See here](https://nus-cs2113-ay2223s2.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-magic-numbers)

sevenseasofbri commented on src/main/java/com/ethanyidong/bunny/BunnySession.java:
```diff
@@ -0,0 +1,68 @@
+import java.util.Arrays;
+
+public class BunnySession {
+    private final static String DIVIDER = "____________________________________________________________";
+    private final static String GLOBAL_INDENTATION = "\t";
+    private final static String MESSAGE_INDENTATION = " ";
```
Good use of constants

sevenseasofbri commented on src/main/java/com/ethanyidong/bunny/cmd/ExecutableCommand.java:
```diff
@@ -0,0 +1,60 @@
+        return newExecutableCommand;
+    }
+
+    protected CommandValidator[] validators() {
```
The name of a function should describe what it does. As such, `validators` does not give the reader a clear picture of what the function is supposed to do/return. Perhaps consider renaming it to `getValidators` or something along those lines. [See here](https://nus-cs2113-ay2223s2.github.io/website/se-book-adapted/chapters/codeQuality.html#use-name-to-explain)

sevenseasofbri commented on src/main/java/com/ethanyidong/bunny/cmd/TokenizedCommand.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    private static HashMap<String, String> extractFlagArguments(String[] positionalAndFlagArguments) {
+        HashMap<String, String> ret = new HashMap<>();
```
Variable name `ret` may not be descriptive enough to explain what this hashmap does. Consider renaming it. [See here](https://nus-cs2113-ay2223s2.github.io/website/se-book-adapted/chapters/codeQuality.html#use-name-to-explain)

sevenseasofbri commented on src/main/java/com/ethanyidong/bunny/cmd/ListCommand.java:
```diff
@@ -0,0 +1,19 @@
+import java.util.ArrayList;
+
+public class ListCommand extends ExecutableCommand {
+    public void execute(BunnySession bunny) {
```
Consider avoiding deep nesting to improve readability. [See here](https://nus-cs2113-ay2223s2.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-deep-nesting)

sevenseasofbri reviewed (commented):
Good job Ethan! 💯 You have mostly followed the coding standard conventions, the readability is good, but can be improved further. I have added comments where necessary but if you think my comment applies to another segment of code I haven't commented on, please exercise the change there too.

EthanYidong committed:
Refractor code to extract save file, UI, and state functionality

Create BunnyTaskList to handle state
Create BunnyUI to handle user interaction
Create BunnyStorage to handle save file
Modified command classes to utilize new abstractions

EthanYidong committed:
Implement recognizing and formatting ISO dates

Create FlexibleDate class to represent dates which may or may not be formatted
Update Deadline and Event classes to use FlexibleDate instead of String to store dates
Update test to test formatted dates

EthanYidong committed:
Implement find command

Create FindCommand class
Add find command to ExecutableCommand list

EthanYidong committed:
Add documentation to most methods

Refractor: reduce visibility for some public methods which did not need to be public
Refractor: delete unused delete methods

EthanYidong committed:
Merge pull request #1 from EthanYidong/branch-Level-8

Implement recognizing and formatting ISO dates

EthanYidong committed:
Merge pull request #2 from EthanYidong/branch-Level-9

Implement find command

EthanYidong committed:
Merge branch 'master' into branch-A-JavaDoc

EthanYidong committed:
Merge pull request #3 from EthanYidong/branch-A-JavaDoc

Add documentation to most methods

EthanYidong committed:
Add documentation to the FindCommand class

Fixed typo in ListCommand documentation
Small refractor to the visibility of methods in the FindCommand class

EthanYidong committed:
Update user guide to list all current features and commands

EthanYidong committed:
Add links to user guide

EthanYidong committed:
Refractor for code quality

Ran IntelliJ autoformatting on all code
Rename variables to align with coding standards
Extract output strings and output errors to constant format strings in UI class

EthanYidong committed:
Continue running if writing to save file fails

EthanYidong committed:
Change file storage to be relative to JAR file

EthanYidong committed:
Fix bug where a command with positional argument / would crash the program
Fix bug where state is not saved if the buffer is not flushed

EthanYidong committed:
Swap saving and running commands to prevent unusable states in the presense of bugs
